### PR TITLE
Datasource moved and refactored

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description
From previous experiences with the datasource controller, we didnt utilize the controller and just wrapped the whole thing in React-Query. Im having doubts about the approach of the datasource controller. I dont like having a dependency like React-Query that cant be used before we reach the react-workspace wrapper. However I dont want to write a similar package to React-Query just in pure JS. 

I'm not sure we need a datasource controller. We can maybe introduce a React specific one, and utilize React-Query for it?

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Checklist

- [x] I have performed a self-review of my own code.
- [ ] I have linked my DevOps task using the AB# tag.
- [x] My code is easy to read.
- [x] Documentation and comments is added where needed.
- [x] My code is covered by tests.

### CODE REVIEW

File: packages/data-source/src/lib/classes/dataSourceController.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            if (!preventCallbacks) {
+                this.notifyDataChanged();
+            }
+            return data;
+
+        }catch(e: unknown){
+            this.throwError(e as TError)
+        }
+
+        return [];
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/classes/dataSourceController.ts:41]: return data??

File: packages/data-source/src/lib/classes/dataSourceController.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+
+import { defaultResponseParser, generateUniqueId } from '../functions';
+import { FetchResponseAsync, OnDataChangedCallback, OnErrorCallback, ResponseParserAsync } from '../types';
+import { Callback, OnCallbackSet } from '../types/callback';
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/classes/dataSourceController.ts:4]: import!

File: packages/data-source/src/lib/classes/dataSourceController.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    remove = (): void => {
+        this.data = [];
+        this.notifyDataChanged();
+    }
+
+    /**
+     * Register a callback to be called whenever data changes
+     * @param cb 
+     * @returns
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/classes/dataSourceController.ts:71]: comments

File: packages/data-source/src/lib/classes/dataSourceController.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            unSubscribe: () => {
+                this.onDataChangedCallbacks = this.onDataChangedCallbacks.filter(
+                    (callback) => callback.id !== id
+                );
+            },
+        };
+    };
+    
+
+    onErrorThrown = (cb: OnErrorCallback<TData,TError>): OnCallbackSet => {
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/classes/dataSourceController.ts:87]: TSDoc

File: packages/data-source/src/lib/functions/defaultResponseParser.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Default response parser if no custom parser is provided.
+ * @param res
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/functions/defaultResponseParser.ts:3]: fix tsdoc

File: packages/data-source/src/lib/classes/dataSourceController.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
+
+import { defaultResponseParser, generateUniqueId } from '../functions';
+import { FetchResponseAsync, OnDataChangedCallback, OnErrorCallback, ResponseParserAsync } from '../types';
+import { Callback, OnCallbackSet } from '../types/callback';
+
+export class DataSourceController<TData, TError = unknown> {
+    /** Function that returns the api call promise */
+    fetchResponseAsync: FetchResponseAsync;
+    /** Function that parses the response to correct format, defaults to just parsing the raw response */
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/classes/dataSourceController.ts:9]: merge

File: packages/data-source/src/lib/test/onDataChanged.spec.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { DataSourceController } from "../classes";
+
+async function defaultFetch(){
+    return {} as any
+}
+const mocFunction = jest.fn();
+
+describe("OnDataChangedTesting", () => {
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onDataChanged.spec.ts:8]: Elaborate

File: packages/data-source/src/lib/test/onDataChanged.spec.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    });
+    it("Should not run onDataChanged when preventCallbacks is set to true", async () => {
+        const {unSubscribe} = controller.onDataChanged(() => {
+            throw "Function should not have been called, fetchData is ignoring the preventCallbacks flag"
+        })
+        await controller.fetchData(true);
+        unSubscribe();
+    })
+
+
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onDataChanged.spec.ts:33]: Remove ws

File: packages/data-source/src/lib/test/onDataChanged.spec.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        await controller.fetchData();
+        expect(mocFunction).toBeCalledTimes(1)
+        unSubscribe();
+    });
+    it("Should not run onDataChanged when preventCallbacks is set to true", async () => {
+        const {unSubscribe} = controller.onDataChanged(() => {
+            throw "Function should not have been called, fetchData is ignoring the preventCallbacks flag"
+        })
+        await controller.fetchData(true);
+        unSubscribe();
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onDataChanged.spec.ts:30]: No expects?

File: packages/data-source/src/lib/test/onDataChanged.spec.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    it("Should run onDataChangedCallback", async () => {
+        const {unSubscribe} = controller.onDataChanged((s) => {
+            mocFunction();
+            expect(s.length).toBe(3)
+        })
+        await controller.fetchData();
+        expect(mocFunction).toBeCalledTimes(1)
+        unSubscribe();
+    });
+    it("Should not run onDataChanged when preventCallbacks is set to true", async () => {
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onDataChanged.spec.ts:25]: add comment to explain that if the function is called the test will fail

File: packages/data-source/src/lib/test/onError.spec.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+const errorMessage = "Some error";
+
+
+async function defaultFetch(): Promise<Response>{
+    throw "Some error"
+}
+
+
+describe("OnDataChangedTesting", () => {
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onError.spec.ts:11]: Describe properly!

File: packages/data-source/src/lib/test/onError.spec.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { DataSourceController } from "../classes";
+
+const errorMessage = "Some error";
+
+
+async function defaultFetch(): Promise<Response>{
+    throw "Some error"
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/test/onError.spec.ts:7]: use errorMessage

File: packages/data-source/src/lib/types/events.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+import { DataSourceController } from "../classes";
+
+/**
+ * On data changed event callback function
+ */
+export type OnDataChangedCallback<T, TError> = (data: T[], controller: DataSourceController<T,TError>) => void;
+export type OnErrorCallback<TData, TError> = (error: TError, controller: DataSourceController<TData, TError>) => void;
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/types/events.ts:7]: comment

File: packages/data-source/src/lib/types/events.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import { DataSourceController } from "../classes";
+
+/**
+ * On data changed event callback function
```
Comment (Gustav-Eikaas) [inline packages/data-source/src/lib/types/events.ts:4]: better comment

**Review by jaysencpp — APPROVED** [review]

**Pull request merged**